File: Python/IndentationErrorListener.py
##
## Project      : an ErrorListener class to collect and display indentation errors
##

from antlr4 import *
from antlr4.error.ErrorListener import ErrorListener

class IndentationErrorListener(ErrorListener):
    _errorList = []

    @staticmethod
    def lexerError(msg: str):
        IndentationErrorListener._errorList.append(msg)

    def syntaxError(self, recognizer
                  , offendingSymbol
                  , line: int, charPositionInLine: int
                  , msg: str
                  , e: RecognitionException):

        startOfMessage = " line " + str(line) + ":\t "

##      ************************************************************
##      *** Not exact matches! This is only for a demonstration. ***
##      ************************************************************
        if msg.startswith("missing INDENT at "):
            IndentationErrorListener._errorList.append(startOfMessage + "expected an indented block")
        elif msg.startswith("mismatched input '<INDENT>") or \
             msg.startswith("extraneous input '<INDENT>"):

            IndentationErrorListener._errorList.append(startOfMessage + "unexpected indent")

    @staticmethod
    def displayErrors():
       if len(IndentationErrorListener._errorList) > 0:
            errors = "\nINDENTATION ERROR:\n" + "\n".join(IndentationErrorListener._errorList)
            print(errors)

File: Java/IndentationErrorListener.java
/*
 * Project      : an ErrorListener class to collect and display indentation errors
 */

import org.antlr.v4.runtime.*;

import java.util.ArrayList;
import java.util.List;

public class IndentationErrorListener extends BaseErrorListener {
    private static final List<String> _errorList = new ArrayList<>();

    public static void lexerError(String msg) {
        _errorList.add(msg);
    }

    @Override
    public void syntaxError(Recognizer<?, ?> recognizer,
                            Object offendingSymbol,
                            int line, int charPositionInLine,
                            String msg,
                            RecognitionException e) {

        final String startOfMessage = " line " + line + ":\t ";

        // ************************************************************
        // *** Not exact matches! This is only for a demonstration. ***
        // ************************************************************
        if (msg.startsWith("missing INDENT at ")) {
            _errorList.add(startOfMessage + "expected an indented block");
        } else if (msg.startsWith("mismatched input '<INDENT>") ||
                   msg.startsWith("extraneous input '<INDENT>")) {

            _errorList.add(startOfMessage + "unexpected indent");
        }
    }

    public static void displayErrors() {
        if (_errorList.size() > 0) {
            String errors = "\nINDENTATION ERROR:\n" + String.join("\n", _errorList);
            System.err.println(errors);
        }
    }
}
